Add active users prometheus metrics

yuvipanda opened the pull request:
These are *extremely useful* for people advocating for more resources for their JupyterHubs, but a little difficult to calculate without a full scale log ingestion and analytics pipeline (such as ELK or equivalent). However, these are easy to calculate on the JupyterHub side at any given instance - these are fairly quick SQL queries. Prometheus can capture and store this as a timeseries, and provide valuble advocacy data that is hard to get otherwise.

This turns the metrics on by default, but only updates them every hour - which seems fine for metrics that don't change that often. This should reduce performance impact. Admins can also turn this off if needed.

I've had to implement this in many different ways in many different contexts, and it's also important to be able to *trust* these. Getting this from other grafana data can be tricky to validate - we had an experimental one in our grafana dashboards at some point in the past that we had to kill due to it being hard to validate (https://github.com/jupyterhub/grafana-dashboards/pull/45). This metric will provide an authoritative source of truth for this.

There is a `period` parameter to `active_users` that provides the time period for which the active users are calculated. It provides '24h', '7d', '30d' - accurate, matches other time descriptors used in prometheus, and far more accurate than 'daily' or 'monthly'.

Ref https://github.com/2i2c-org/infrastructure/issues/1888

yuvipanda commented:
From looking at the lastactivity handler code, it looks like the user objects last_activity field is updated anytime last_activity is updated for any of its servers, which matches what I expect.

minrk commented:
Makes sense! Do you think 1d and 30d are standard enough intervals that we shouldn't provide config to allow deployments to _specify_ the intervals (e.g. 7d)?

consideRatio commented:
> Do you think 1d and 30d are standard enough intervals that we shouldn't provide config to allow deployments to specify the intervals (e.g. 7d)?

Hmm hmmm, I think if we opt to make it configurable, we would make it a bit harder to name the metrics etc and perhaps provide standarized jupyterhub dashboards. Is it worth providing this customization if it means providing a reliable standard solution harder?

manics reviewed (commented):
> provide config to allow deployments to specify the intervals (e.g. 7d)

That was my first thought too! Could make it take an array of intervals (in hours, or maybe seconds, so people can have more fine grained usage? Useful for analysing use over a working day)

And the metric could be something like `jupyterhub_active_users_per_$N`?

minrk commented:
> we would make it a bit harder to name the metrics

If we e.g. limit it to a number of days, we could name them like:

```
active_users_last_{days}d
```

where the number of days is in the metric name. Then the dashboard would only have to assume that one or more values for days have been chosen.

Alternately, we could us a single metric and store the interval as a _label_:

```
active_users_last_days{days=30}
```

which could then probably make a single chart with a line for each interval, no matter what intervals are collected. I _think_ that might be less efficient for prometheus? I'm not sure.


Because of this complexity, I don't think we should add configuration for it if it's not likely folks want it (we don't have any other configurably-defined metrics). I just wanted to get folks' opinions on how likely we think that is, as weighed against the complexity of dealing with it.

Another possible criticism of this metric: it won't work for BinderHub or any deployment that deletes inactive users, since it only tells you how many users _still in the database_ who were active in that interval. To allow counting 'deleted' users, we'd need trash-like mechanism for deleting users: mark users as deleted to remove them from _most_ API endpoints, but actually remove from the db after some interval, e.g. 31d.

yuvipanda commented:
I would prefer to not make this configurable. I think daily and monthly is good enough for most use cases. We can add a *yearly* later if anyone wants, but IMO that's not super common.

I think the TODO items here are:

1. Make sure to use utc here
2. Add a test

I don't think we should make this super configurable (lots of complexity!) or try to solve *all* cases (like binder) here.

yuvipanda commented on jupyterhub/metrics.py:
```diff
@@ -200,3 +214,74 @@ def prometheus_log_method(handler):
+        if self.active_users_metrics_enabled:
+            # Setup periodic refresh of the metric
+            pc = PeriodicCallback(
+                self.update, self.active_users_metrics_update_period * 1000, jitter=0.01
```
@minrk in our case, there's only one process hitting the db so probably not. However, I think all periodic callbacks should have jitter just in case anyway, so I kept it in.

consideRatio commented:
I lack experience here, but I'm considering the difference between multiple metrics like `active_users_last_1day` `active_users_last_30days` and what @minrk considered with a label (with a predefined set of values - 1, 30).

> Alternately, we could us a single metric and store the interval as a _label_:
> 
> ```
> active_users_last_days{days=30}
> ```
> 
> which could then probably make a single chart with a line for each interval, no matter what intervals are collected. I _think_ that might be less efficient for prometheus? I'm not sure.

Hoping its reasonable to use a label like that to avoid defining multiple metric names, thinking it will be easier to consume in jupyterhub/grafana-dashboards.

minrk commented:
Makes sense! I also think we shouldn't implement the rigorous 'monthly active users' metric here, I was trying to note a case where expectations from the metric name/description might not match the behavior.

If we're not making it configurable, does 7d make sense? I'd usually use a multiple of weeks for a metric like this, since it captures the natural user activity cycle. But if nobody would use it, we should leave it out.

consideRatio commented:
pre-commit.ci run

consideRatio commented:
Reopened to get tests against latest main branch, which should make selenium test not fail.

yuvipanda commented:
@minrk @consideRatio @manics so I've renamed this to `active_users`, and added a `period` label that can take one of `24h`, `7d`, `30d` as labels. This is more accurate than 'daily' and 'monthly', and adds the '7d' as well. Thank you for the valuble input :)

If this design looks ok, I can work on adding tests. Here's an example of how the metrics are:

```
# HELP jupyterhub_active_users number of users who were active in the given time period
# TYPE jupyterhub_active_users gauge
jupyterhub_active_users{period="24h"} 1.0
jupyterhub_active_users{period="7d"} 494.0
jupyterhub_active_users{period="30d"} 3140.0
```

manics commented:
```
jupyterhub_active_users{period="24h"} 1.0
jupyterhub_active_users{period="7d"} 494.0
jupyterhub_active_users{period="30d"} 3140.0
```
I like this, and it's easily extensible in future 🌞

consideRatio commented:
> If this design looks ok, I can work on adding tests. Here's an example of how the metrics are:
> 
> ```
> # HELP jupyterhub_active_users number of users who were active in the given time period
> # TYPE jupyterhub_active_users gauge
> jupyterhub_active_users{period="24h"} 1.0
> jupyterhub_active_users{period="7d"} 494.0
> jupyterhub_active_users{period="30d"} 3140.0
> ```

@yuvipanda looks great!!! :tada:

minrk commented:
Sounds great, @yuvipanda!

consideRatio commented:
@yuvipanda you have not pushed recent changes but I suspect you may have intended to do so

yuvipanda commented:
@consideRatio whoops good catch, fixed!

consideRatio commented on jupyterhub/metrics.py:
```diff
@@ -44,6 +50,16 @@
+    ['period'],
+)
+
+MONTHLY_ACTIVE_USERS = Gauge(
+    'jupyterhub_monthly_active_users', 'number of users who were active in the last 30d'
+)
+
```
```suggestion
```

consideRatio commented on jupyterhub/metrics.py:
```diff
@@ -200,3 +230,69 @@ def prometheus_log_method(handler):
+
+    db = Any(help="SQLAlchemy db to use for performing queries")
+
+    def update(self):
```
It seems fine that the class is designed ahead of time to be general purpose for other metrics to updated periodically, but I think the update function needs to be specific to active_users. So `update_active_users` for example?

The `start` function is called, and checks if active_users_metrics_enabled is truthy, and then a periodic callback is scheduled to run this `update` function. So that makes me think it should be `update_active_users` rather than a general `update` function.

minrk replied:
I'm torn! I think `update` is the right name. This is a generic class for periodically updating metrics. Active users happens to be the only kind we have right now, so that's the only checks we have, but adding more wouldn't change what the update method should be called. It might move exactly where the `enabled` checks happen, though.

manics replied:
It sounds like this class is trying to be both an Abstract class (`start()` and `update()` methods) and a concrete implementation (defines `active_users_metrics_enabled` and `active_users_metrics_update_period` which will be inherited by any subclasses of `PeriodicMetricsCollector` even if it's used for some other metrics. Would it make sense to separate out the Abstract class, or is that over engineering things?

consideRatio replied:
Update can be the name, but then the active_users_metrics_update_period is a bit off.

Since its now a class designed for future metrics to be updated periodically, i think it should be natural on how to go about it. But, there can only be ome update period as its done now right - that is why either the update name or the active_users_metrics_update_period doesnt feel right.

consideRatio replied:
In other words, pretend another metrics is added to use this class. How should it be added? Maybe it was meant as a base class, and when a new class is added we need to extract active_users_metric, or maybe it was meant as a single class with one update interval, or a class that should handle multiple update intervals. A decision should be made now, or a comment indicating that whoever adds something in the future can do as they think makes sense.

minrk replied:
> either the update name or the active_users_metrics_update_period doesnt feel right.

👍 I agree, they should probably match. Since @yuvipanda named the config value with `active_users`, I made the method name match, so presumably adding another periodically updated metric will mean a new config for enabled/interval and a new update_$metrics method.

> It sounds like this class is trying to be both an Abstract class

@manics I don't think it's meant to be an abstract class at all, I think it's meant to be the one and only periodic metrics updater instance. So if we have other periodically updated metrics, they would be added to the update in this class, not a new class. I think the only question is whether the period should be shared across all metrics or per metric group, and naming follows from that. Naming of methods is a private implementation detail and not something I'd worry about, but naming of config is a user-facing API we need to consider because it's a pain to change.

yuvipanda replied:
Yep, it was meant to be sort of a singleton, rather than an abstract class. I am extremely happy with the state it was merged in! If we add more of these metrics in the future, i think it makes sense to add individual traitlet knobs for those, and we can even coalesce periodcallbacks if needed.

Thank you so much @minrk for taking this to completion.

minrk commented:
@yuvipanda I'd like to make 3.1 this week, do you think you'll have a chance to finish this? If not, I can have a look at the tests.

yuvipanda commented:
@minrk i don't think I can add tests this week, would *LOVE* if you could finish it up!!!!

minrk commented:
I added the tests and all looks good. One outstanding question to @consideRatio above, then happy to merge.

consideRatio reviewed (approved):
Wieee looks great!

yuvipanda commented:
Thank you so much for taking this through, @minrk!!!

yuvipanda commented:
https://github.com/jupyter-server/jupyter_server/pull/1471 adds a similar metric for jupyter_server!
